### PR TITLE
[ceres] update to 2.2.0

### LINKED ISSUE #34483: [ceres-solver] update to 2.2.0
### Library name

ceres-solver

### New version number

2.2.0

### Other information that may be useful (release notes, etc...)

Ceres-solver http://ceres-solver.org/ just got updated to 2.2.0. Release notes: http://ceres-solver.org/version_history.html

### PR DESCRIPTION
**Pull request by Cheney-W**
Fixes https://github.com/microsoft/vcpkg/issues/34483

According to https://github.com/ceres-solver/ceres-solver/commit/2335b5b4b7a4703ca9458aa275ca878945763a34, remove feature `cxsparse`.
According to https://github.com/ceres-solver/ceres-solver/commit/0c88301e6658f9c52e54621d45e333dce35c09ec, add dependency `metis` for feature `eigensparse`.
Modify `0003_fix_exported_ceres_config.patch` to add the required dependencies.
Modify `find-package-required.patch` to use the ports provided by vcpkg.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] ~The "supports" clause reflects platforms that may be fixed by this new version~
- [ ] ~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~
- [ ] ~Any patches that are no longer applied are deleted from the port's directory.~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/ceres/find-package-required.patch [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
-@@ -266,7 +266,8 @@ if (SUITESPARSE)
+@@ -292,27 +292,26 @@ if (SUITESPARSE)
    # built with SuiteSparse support.
  
    # Check for SuiteSparse and dependencies.
--  find_package(SuiteSparse 4.0 COMPONENTS CHOLMOD SPQR)
+-  find_package(SuiteSparse 4.5.6 COMPONENTS CHOLMOD SPQR
+-    OPTIONAL_COMPONENTS Partition)
+-  if (SuiteSparse_FOUND)
 +  find_package(suitesparse CONFIG REQUIRED)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:21]:
```suggestion
+  find_package(SuiteSparse NAMES suitesparse REQUIRED)
```

File: ports/ceres/find-package-required.patch [hunk 2]
```diff
@@ -1,27 +1,28 @@
...
-+  set(SuiteSparse_FOUND 1)
-   if (SuiteSparse_FOUND)
++  if (suitesparse_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:22]:
```suggestion
+  if (SuiteSparse_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 3]
```diff
@@ -31,33 +32,50 @@ index 2d241fe..b79454d 100644
                      ${Ceres_SOURCE_DIR}/cmake/FindMETIS.cmake
                DESTINATION ${RELATIVE_CMAKECONFIG_INSTALL_DIR})
 -    endif (SuiteSparse_NO_CMAKE OR NOT SuiteSparse_DIR)
+-  else (SuiteSparse_FOUND)
 +    endif ()
-   else (SuiteSparse_FOUND)
++  else (suitesparse_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:37]:
```suggestion
+  else (SuiteSparse_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 4]
```diff
@@ -31,33 +32,50 @@ index 2d241fe..b79454d 100644
...
-     set(CXSparse_DEPENDENCY "find_dependency(CXSparse ${CXSparse_VERSION})")
-     # By default, if CXSparse and all dependencies are found, Ceres is
-@@ -302,10 +303,10 @@ if (CXSPARSE)
-     message("-- Found CXSparse version: ${CXSparse_VERSION}, "
-       "building with CXSparse.")
+       "SuiteSparse support.")
+     update_cache_variable(SUITESPARSE OFF)
+     list(APPEND CERES_COMPILE_OPTIONS CERES_NO_SUITESPARSE)
+-  endif (SuiteSparse_FOUND)
++  endif (suitesparse_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:44]:
```suggestion
+  endif (SuiteSparse_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 5]
```diff
@@ -31,33 +32,50 @@ index 2d241fe..b79454d 100644
...
+   list(APPEND CERES_COMPILE_OPTIONS CERES_NO_SUITESPARSE)
+@@ -323,23 +322,23 @@ if (NOT SuiteSparse_Partition_FOUND)
+ endif (NOT SuiteSparse_Partition_FOUND)
  
--    if (CXSparse_NO_CMAKE OR NOT CXSparse_DIR)
+ if (EIGENMETIS)
+-  find_package (METIS)
+-  if (METIS_FOUND)
++  find_package(metis CONFIG REQUIRED)
++  if (metis_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:55]:
```suggestion
+  find_package(METIS NAMES metis REQUIRED)
+  if (METIS_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 6]
```diff
@@ -31,33 +32,50 @@ index 2d241fe..b79454d 100644
...
--    endif (CXSparse_NO_CMAKE OR NOT CXSparse_DIR)
-+    endif ()
-   else (CXSparse_FOUND)
-     # Disable use of CXSparse if it cannot be found and continue.
-     message("-- Did not find CXSparse, Building without CXSparse.")
-@@ -367,9 +368,10 @@ endif()
+-    endif (NOT METIS_DIR)
+-  else (METIS_FOUND)
++    endif (0)
++  else (metis_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:69]:
```suggestion
+  else (METIS_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 7]
```diff
@@ -31,33 +32,50 @@ index 2d241fe..b79454d 100644
...
+     message("-- Did not find METIS, disabling Eigen METIS support.")
+     update_cache_variable(EIGENMETIS OFF)
+     list (APPEND CERES_COMPILE_OPTIONS CERES_NO_EIGEN_METIS)
+-  endif (METIS_FOUND)
++  endif (metis_FOUND)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:74]:
```suggestion
+  endif (METIS_FOUND)
```

File: ports/ceres/find-package-required.patch [hunk 8]
```diff
@@ -66,47 +84,48 @@ index 2d241fe..b79454d 100644
    if (gflags_FOUND)
 -    if (TARGET gflags)
 +    if (TARGET gflags::gflags)
-+      add_library(gflags ALIAS gflags::gflags)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:69]: Keep this ...

File: ports/ceres/find-package-required.patch [hunk 9]
```diff
@@ -66,47 +84,48 @@ index 2d241fe..b79454d 100644
...
++  list(APPEND CERES_LIBRARY_PRIVATE_DEPENDENCIES metis)
+ endif (SuiteSparse_Partition_FOUND OR EIGENMETIS)
  
  if (ACCELERATESPARSE AND AccelerateSparse_FOUND)
+@@ -396,7 +396,7 @@ if (BUILD_TESTING AND GFLAGS)
+   if (CMAKE_SYSTEM_NAME MATCHES "QNX")
+     target_link_libraries(gtest PUBLIC regex)
+   endif()
+-  target_link_libraries(gtest PRIVATE Ceres::ceres gflags)
++  target_link_libraries(gtest PRIVATE Ceres::ceres gflags::gflags)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:119]: ... and you don't have to patch that ...

File: ports/ceres/find-package-required.patch [hunk 10]
```diff
@@ -66,47 +84,48 @@ index 2d241fe..b79454d 100644
...
+ 
+   add_library(test_util STATIC
+               evaluator_test_utils.cc
+@@ -404,7 +404,7 @@ if (BUILD_TESTING AND GFLAGS)
+               test_util.cc)
+ 
+   target_include_directories(test_util PUBLIC ${Ceres_SOURCE_DIR}/internal)
+-  target_link_libraries (test_util PUBLIC ceres_static gflags gtest)
++  target_link_libraries (test_util PUBLIC ceres_static gflags::gflags gtest)
```
Comment (dg0yt) [inline ports/ceres/find-package-required.patch:128]: ... and that.

**Review by dg0yt — COMMENTED** [review]
Make good use of `find_package(... NAMES ...)` and `add_library(...  ALIAS ...)` to trim the patch size.

File: ports/ceres/find-package-required.patch (hunk 8 above)
Comment (dg0yt) [reply ports/ceres/find-package-required.patch:69]: IMO this error is a vcpkg bug:  Abuse of `vcpkg-cmake-wrapper.cmake` in port gflags.

Pristine upstream behaviour of `find_package(gflags)`:  
The use of a namespace is controlled by the `GFLAGS_USE_TARGET_NAMESPACE` input variable, 
and the default value is `OFF`.  
https://github.com/gflags/gflags/blob/master/cmake/config.cmake.in#L10-L12

vcpkg behaviour:
The wrapper unconditionally sets `GFLAGS_USE_TARGET_NAMESPACE` to `ON`, 
https://github.com/microsoft/vcpkg/blob/0c20b2a97c390e106150837042d921b0939e7ecb/ports/gflags/vcpkg-cmake-wrapper.cmake#L1,
and in addition, it creates the non-prefixed targets,
https://github.com/microsoft/vcpkg/blob/0c20b2a97c390e106150837042d921b0939e7ecb/ports/gflags/vcpkg-cmake-wrapper.cmake#L5-L10.

So with plain `find_package(gflags)`, we get namespaced targets which we didn't expect,
and with a user-side `set(GFLAGS_USE_TARGET_NAMESPACE ON)`, we get non-namespaced targets which we didn't expect.

Comment (dg0yt) [reply ports/ceres/find-package-required.patch:69]: The conclusion is to trim the patch. Upstream has it right.
~~~
find_package(gflags CONFIG)
target_link_libraries(foo PRIVATE gflags)
~~~

**Comment by patrikhuber** [comment]
> According to https://github.com/ceres-solver/ceres-solver/commit/0c88301e6658f9c52e54621d45e333dce35c09ec, add dependency metis for feature eigensparse.

If I understood you correctly, you want to add a dependency on metis whenever the feature ceres[eigensparse] is requested? As far as I understand https://github.com/ceres-solver/ceres-solver/commit/0c88301e6658f9c52e54621d45e333dce35c09ec, metis is *not* a dependency of eigensparse, it's an optional extra - but I'm not 100% sure. If that is the case, then please don't pull in metis as a dependency whenever someone requests ceres[eigensparse].

**Comment by hesmar** [comment]
What is the status of this PR? Is anyone still working on it? I would like to create another patch for Ceres which will be based on 2.2.0. So it would be nice to have this in master.

**Comment by patrikhuber** [comment]
It would be really great to have Ceres 2.2.0 available in vcpkg, it's been out for quite a few months now.

**Comment by hesmar** [comment]
@Cheney-W Are you still working on it?

**Comment by moolawooda** [comment]
So, is there any progress at the moment?

**Comment by saulthu** [comment]
Bump🤜

**Comment by saulthu** [comment]
@Cheney-W Any update on why this was closed? (Thanks for the efforts BTW!)

**Comment by valgur** [comment]
Just FYI, I intend to re-open a v2.2.0 PR after #39297 is merged.

**Comment by Cheney-W** [comment]
I encountered some compilation issues during the update, and I'm not sure how to resolve them yet. To avoid blocking others from updating this port, I've decided to close this PR for now. I will carefully investigate the errors on my local machine, and once I resolve them, if no one else has updated this port, I will reopen my PR.

**Comment by ahojnnes** [comment]
Another attempt here: https://github.com/microsoft/vcpkg/pull/42475 - getting close.

**Pull request closed without merging**